src/main/java/com/scalar/cassy/transferer/AzureBlobFileUploader.java

```diff
@@ -40,57 +46,83 @@ public AzureBlobFileUploader(BlobContainerAsyncClient blobContainerClient) {
 
     logger.info("Uploading " + file);
-    return blobContainerClient
-        .getBlobAsyncClient(key)
-        .uploadFromFile(file.toString(), true)
-        .toFuture();
+    return executorService.submit(
+        () -> {
+          try {
+            try (InputStream inputStream = new FileInputStream(file.toString())) {
+              blobContainerClient
+                  .getBlobClient(key)
+                  .upload(inputStream, new File(file.toString()).length(), true);
+            }
+          } catch (IOException e) {
+            throw new FileTransferException(e);
+          }
+
+          logger.info("Upload file succeeded : " + file.toString());
+          return null;
+        });
   }
 
   @Override
   public void upload(List<Path> files, BackupConfig config) {
     Path dataDir = Paths.get(config.getDataDir());
-    List<Mono<Void>> toBeUploaded = new ArrayList<>();
+    List<Future> uploads = new ArrayList<>();
     for (Path filePath : files) {
       Path relative = dataDir.relativize(filePath);
       String key = BackupPath.create(config, relative.toString());
       if (requiresUpload(key, filePath)) {
         logger.info("Uploading " + filePath);
-        toBeUploaded.add(
-            blobContainerClient
-                .getBlobAsyncClient(key)
-                .uploadFromFile(filePath.toString(), true)
-                .doOnSuccess(blobProperties -> logger.info("Upload succeeded : " + filePath))
-                .doOnError(
-                    error -> {
-                      throw new FileTransferException("Upload failed : " + filePath, error);
-                    }));
+        uploads.add(
+            executorService.submit(
+                () -> {
+                  try {
+                    try (InputStream inputStream = readStream(filePath)) {
+                      blobContainerClient
+                          .getBlobClient(key)
+                          .upload(inputStream, new File(filePath.toString()).length(), true);
+                    }
+                  } catch (IOException e) {
+                    throw new FileTransferException(e);
+                  }
+                  logger.info("Upload file succeeded : " + filePath.toString());
+                }));
       } else {
         logger.info(filePath + " has been already uploaded.");
       }
 
-      if (toBeUploaded.size() >= ASYNC_FILE_UPLOAD_LIMIT) {
-        // Start uploading files asynchronously and wait for them to complete
-        Mono.when(toBeUploaded).block();
-        toBeUploaded.clear();
+      if (uploads.size() >= ASYNC_FILE_UPLOAD_LIMIT) {
+        waitForAsyncFileUpload(uploads);
+        uploads.clear();
       }
     }
 
-    if (toBeUploaded.size() > 0) {
-      // Start uploading files asynchronously and wait for them to complete
-      Mono.when(toBeUploaded).block();
+    if (uploads.size() > 0) {
+      waitForAsyncFileUpload(uploads);
     }
   }
 
+  private void waitForAsyncFileUpload(List<Future> uploads) {
```

Comment:
```suggestion
  private void waitForAsyncFileUpload(List<Future<Void>> uploads) {
```

Reply:
Thank you for your suggestion.
Suggested changes integrated f8d33d6